drivers/crc/crc_stm32.c

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright (c) 2024 Brill Power Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(crc_stm32, CONFIG_CRC_HW_LOG_LEVEL);
+
+#include <errno.h>
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+#include <zephyr/drivers/crc.h>
+#include <zephyr/sys/byteorder.h>
+
+#include <soc.h>
+#include <stm32_ll_crc.h>
+
+#define DT_DRV_COMPAT st_stm32_crc
+
+struct crc_stm32_cfg {
+	CRC_TypeDef *base;
+	struct stm32_pclken pclken;
+};
+
+struct crc_stm32_data {
+	struct k_mutex mutex;
+	k_timeout_t timeout;
+};
+
+static int crc_lock(const struct device *dev, k_timeout_t timeout)
+{
+	struct crc_stm32_data *data = (struct crc_stm32_data *)dev->data;
+
+	return k_mutex_lock(&data->mutex, timeout);
+}
+
+static int crc_unlock(const struct device *dev)
+{
+	struct crc_stm32_data *data = (struct crc_stm32_data *)dev->data;
+	struct k_mutex crc_mtx = data->mutex;
+	int res = 0;
+
+	/**
+	 * The lock count can be greater than 1 if the same thread acquires
+	 * a lock multiple times, so we have to iterate until lock_count
+	 * is zero
+	 */
+	while (crc_mtx.lock_count != 0) {
+		res = k_mutex_unlock(&data->mutex);
+
+		if (res != 0) {
+			return res;
+		}
+	}
+
+	return res;
+}
+
+static int crc_stm32_begin(const struct device *dev, struct crc_ctx *ctx)
+{
+	/* Attempt to take mutex lock */
+	const struct crc_stm32_data *data = dev->data;
+	int mtx = crc_lock(dev, data->timeout);
+
+	if (mtx != 0) {
+		return mtx;
+	}
+
+	const struct crc_stm32_cfg *cfg = dev->config;
+
+	LL_CRC_SetInputDataReverseMode(cfg->base, (ctx->flags & CRC_FLAG_REVERSE_INPUT)
+							  ? LL_CRC_INDATA_REVERSE_WORD
+							  : LL_CRC_INDATA_REVERSE_NONE);
+	LL_CRC_SetOutputDataReverseMode(cfg->base, (ctx->flags & CRC_FLAG_REVERSE_OUTPUT)
+							   ? LL_CRC_OUTDATA_REVERSE_BIT
+							   : LL_CRC_INDATA_REVERSE_NONE);
+	LL_CRC_ResetCRCCalculationUnit(cfg->base);
+
+	switch (ctx->type) {
+	case CRC8_CCITT: {
+		uint8_t init_val = (uint8_t)(ctx->initial_value & 0xFF);
+		uint8_t poly = (uint8_t)(ctx->polynomial & 0xFF);
+
+		LL_CRC_SetPolynomialSize(cfg->base, LL_CRC_POLYLENGTH_8B);
+		LL_CRC_SetPolynomialCoef(cfg->base, poly);
+		LL_CRC_SetInitialData(cfg->base, init_val);
+		break;
+	}
+
+	case CRC32_IEEE: {
+		uint32_t init_val = (uint32_t)(ctx->initial_value & 0xFFFFFFFF);
+		uint32_t poly = (uint32_t)(ctx->polynomial & 0xFFFFFFFF);
+
+		LL_CRC_SetPolynomialSize(cfg->base, LL_CRC_POLYLENGTH_32B);
+		LL_CRC_SetPolynomialCoef(cfg->base, poly);
+		LL_CRC_SetInitialData(cfg->base, init_val);
+		break;
+	}
+
+	default:
+		crc_unlock(dev);
+		return -ENOTSUP;
```

Comment:
This locking mechanism looks really flawed, a mutex should be used to protect different API calls, you need another semaphore or something to allow for asynchronous behavior.

Reply:
What is the use case you are envisioning? The CRC peripheral has a single 32-bit register.